### PR TITLE
[Fix](bangc-ops): roiaware add Param Check and use __cn_scalar_sin_*

### PR DESCRIPTION
**Pull request by chqy99**
Thanks for your contribution and we appreciate it a lot. :rocket::rocket:

## 1. Motivation

MMCV中存在隐含的限制条件，在host端新增相应的参数检查。

- dim3结构体，限制了boxes_num和channels大小。

- pts_mask指针的最大偏移值为 boxes_num * pts_num，它们均为int类型。

sin/cos指令精度与竞品不对齐。

nram_max_pts_feature_tmp该内存空间未使用过。

## 2. Modification

a.**mlu-ops/bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp**中 
- boxes_num < 65536
- channels < 65536
- product of boxes_num and pts_num < 2^31

b.**mlu-ops/bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d_union1.mlu**中
- delete unused memory (nram_max_pts_feature_tmp)
- __cn_scalar_sin_* replace std::sin;__cn_scalar_cos_* replace std::cos

c.**mlu-ops/bangc-ops/mlu_op.h**中
-增加参数限制的注释

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- dynamic threshold
  - [x] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [x] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [x] diff4: mlu diff4 <= max(baseline diff4 * 1, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float                   |          |          ||
|Mult-platform test   |MLU370/MLU590    release_test+release_temp        |     all pass     |          |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

case1
|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|baseline|   5575.7 |    |   0.08310758012  |  0.08687124487  |    |   float |   'input_dim': [[128, 7], [16000, 3], [16000, 16]]  |
|modified|  5592.39  |    |   0.08310758012  |  0.08661198522  |    |  float  |   'input_dim': [[128, 7], [16000, 3], [16000, 16]]  |

10个测试样例，性能下降约1%。

Platform：MLU590

case1
|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|baseline|  3608.8  |    |  0.01428432247  | 0.05592440701   |    |  float  |  {'input_dim': [[128, 7], [16000, 3], [16000, 16]]}   |
|modified|   3615.23 |    |  0.01428432247  |  0.0558249406  |    |  float  |  {'input_dim': [[128, 7], [16000, 3], [16000, 16]]}   |

10个测试样例，性能下降约1%。

### 3.4 Summary Analysis

功能流水和性能流水测试均通过。

### CODE REVIEW

File: bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp [hunk 1]
```diff
@@ -209,6 +209,7 @@ mluOpStatus_t MLUOP_WIN_API mluOpRoiawarePool3dForward(
 
   // check tensor dim
   PARAM_CHECK(API, boxes_num > 0);
+  PARAM_CHECK(API, boxes_num < 65536);
```
Comment (mahxn0) [inline bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp:212]: 在limitation和mlu_op.h中增加boxes_num的限制说明．

**Review by mahxn0 — APPROVED** [review]

File: bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp [hunk 2]
```diff
@@ -209,8 +209,10 @@ mluOpStatus_t MLUOP_WIN_API mluOpRoiawarePool3dForward(
...
   PARAM_CHECK(API, pts_num > 0);
   PARAM_CHECK(API, channels > 0);
+  PARAM_CHECK(API, channels < 65536);
```
Comment (PetrelYy) [inline bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp:215]: 用变量来接65536吧，变量名 **_threshold

**Review by PetrelYy — APPROVED** [review]

File: bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp [hunk 3]
```diff
@@ -207,10 +207,16 @@ mluOpStatus_t MLUOP_WIN_API mluOpRoiawarePool3dForward(
   // check other parms : pool_method
   PARAM_CHECK(API, pool_method == 0 || pool_method == 1);
 
+  /* boxes_num or channels is the y- or z-dimension in mmcv(cuda),
+     Maximum y- or z-dimension of a grid of thread blocks
+     should be less than 65536 in cuda. */
+  const int threshold_of_boxes_num_and_channels = 65536;
```
Comment (DanieeelLiu) [inline bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp:213]: 这里建议定义成宏

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -8366,6 +8366,9 @@ mluOpGetRoiawarePool3dForwardWorkspaceSize(mluOpHandle_t handle,
  * - None.
  *
  * @par Scale Limitation
+ * - \b boxes_num should less than 65536.
+ * - \b channels should less than 65536.
```
Comment (duzekunKTH) [inline bangc-ops/mlu_op.h:8370]:
```suggestion
 * - \b boxes_num should be less than 65536.
 * - \b channels should be less than 65536.
```

File: bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp [hunk 5]
```diff
@@ -482,10 +488,17 @@ mluOpStatus_t MLUOP_WIN_API mluOpRoiawarePool3dBackward(
   PARAM_CHECK(API, pool_method == 0 || pool_method == 1);
 
   const int pts_num = grad_in_desc->dims[0];
+
+  /* boxes_num or channels is the y- or z-dimension in mmcv(cuda),
+    Maximum y- or z-dimension of a grid of thread blocks
+    should be less than 65536 in cuda. */
+  const int threshold_of_boxes_num_and_channels = 65536;
```
Comment (DanieeelLiu) [inline bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp:495]: 与上面重复了

File: bangc-ops/mlu_op.h (hunk 4 above)
Comment (duzekunKTH) [inline bangc-ops/mlu_op.h:8370]:
```suggestion
 * - \b boxes_num should be less than 65536.
 * - \b channels should be less than 65536.
```

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -8366,6 +8366,9 @@ mluOpGetRoiawarePool3dForwardWorkspaceSize(mluOpHandle_t handle,
...
+ * - The Product of \b boxes_num and \b pts_num should less than 2G.
```
Comment (duzekunKTH) [inline bangc-ops/mlu_op.h:8371]:
```suggestion
 * - \b boxes_num should be less than 65536.
 * - \b channels should be less than 65536.
 * - The Product of \b boxes_num and \b pts_num should be less than 2G.
```

File: bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d_union1.mlu [hunk 7]
```diff
@@ -321,16 +329,13 @@ __mlu_entry__ void MLUMultiKernelRoiawarePool3dForward(
   int align_max_pts_each_voxel = PAD_UP(max_pts_each_voxel, align_num);
   int nram_channels_limit =
       PAD_DOWN((MAX_NRAM_SIZE - 128 -
-                align_max_pts_each_voxel * (sizeof(int) + sizeof(T))) /
+                align_max_pts_each_voxel * sizeof(int)) /
                    ((align_max_pts_each_voxel + 1) * sizeof(T) + sizeof(int)),
-               align_num);
+                align_num);
```
Comment (DanieeelLiu) [inline bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d_union1.mlu:334]: 这里格式看着有点奇怪，是否用tools/format2goole 刷格式了？

File: bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp [hunk 8]
```diff
@@ -232,12 +238,12 @@ mluOpStatus_t MLUOP_WIN_API mluOpRoiawarePool3dForward(
...
   TENSOR_NUM_CHECK(API, tensor_pooled_features_num, LARGE_TENSOR_NUM, "");
 
-  // check zero element
-  PARAM_CHECK(API, tensor_rois_num > 0);
-  PARAM_CHECK(API, tensor_pts_num > 0);
-  PARAM_CHECK(API, tensor_pts_feature_num > 0);
-  PARAM_CHECK(API, tensor_pts_idx_of_voxels_num > 0);
-  PARAM_CHECK(API, tensor_pooled_features_num > 0);
+  // product of boxes_num and pts_num < 2^31
+  if (boxes_num > INT32_MAX / pts_num) {
```
Comment (duzekunKTH) [inline bangc-ops/kernels/roiaware_pool3d/roiaware_pool3d.cpp:242]:
```suggestion
  size_t product_boxes_pts = (size_t)boxes_num * (size_t)pts_num;
  if (product_boxes_pts >= (size_t)INT32_MAX) {
```
你原本的式子无法排除 boxes_num * pts_num = INT32_MAX 的情况，但改为等号后又会因为 floor 特征出错。

**Review by DanieeelLiu — APPROVED** [review]

**Review by duzekunKTH — APPROVED** [review]

**Pull request merged**